### PR TITLE
Make new adaptive deletion pass actually work

### PR DESCRIPTION
**Pull request by DRMacIver**
Somewhat impressively, the new adaptive deletion pass was literally 100% broken but the tests we had for the pass specifically happened to work given that it now attempted to zero examples, and all of the quality tests the rest of the shrinker was able to pick up the slack.

Um, oops?

### CODE REVIEW

**Comment by DRMacIver** [comment]
Note that I have suppressed some of the quality tests (the poisoned matrix ones). This is because despite having tests for them we can't actually reliably make those pass right now! We were previously getting lucky and now aren't.

Fortunately I have a fix for this in the mega-branch.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def run_step(self):
             i = len(self.current) - 1 - j
             start = self.current
             find_integer(
-                lambda k: k <= i and self.consider(
+                lambda k: k <= i + 1 and self.consider(
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py]: Those aren't the same condition. `k = i + 1` satisfies `k <= i + 1` but `k > i`.

[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py]: ...yep.

**Comment by DRMacIver** [comment]
> Though so did the last one...

This one is at least smaller so hopefully more scope for bugs to stand out (also it's better tested)

File: hypothesis-python/RELEASE.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+RELEASE_TYPE: patch
+
+This release fixes several bugs that were introduced to the shrinker in
+:ref:`3.16.24 <v3.66.24>` which would have caused it to behave significantly
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst]: The link text is still mismatched - 16 != 66

Comment (DRMacIver) [reply hypothesis-python/RELEASE.rst]: Argh. So it is.

File: hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DRMacIver) [reply hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py]: I honestly spent about two minutes staring at your comment going "Is that right? That doesn't *sound* right, but maybe I'm missing something" before coming to that answer. I think we both might be tired. 😆

**Pull request merged**